### PR TITLE
Updates prefetch integration to add "only prefetch link on hover/mouseover/focus" option

### PR DESCRIPTION
**Pull request by kory-smith**
## Changes
This PR adds the ability to prefetch links only on hover by way of an optional intent-selector supplied to the integration configs. By default, it's `intentSelector = 'a[href][rel~="prefetch-intent"]'`, so `<a href="/hover" rel="prefetch-intent">Hover me!</a>` would be prefetched only when one of the following three events occur on it:
- mouseenter
- touchstart
- focus

This new selector is completely optional and doesn't interfere with the current prefetch behavior **at all**. With this new change, you have the ability to only prefetch a link if a user hovers over it with their mouse (or finger), regardless of whether or not it's in the viewport.

It looks like this in the Astro config:
```
	integrations: [
		prefetch({
			intentSelector: 'a[href][rel~="custom-intent"]'
		}),
	],
```

## Motivation
I think that it's sometimes wasteful to preload a link just because it's in the viewport. I think that developers should have the ability to extremely selective in how they spend their users' data. IMO, Remix does it right with their [prefetch=intent](https://remix.run/docs/en/1.14.3/components/link#md-prefetch) option. This is just a copy of that.

## Testing

I added a set of tests for both the default selector and custom selector options. The tests both have prefetch-intent links on the page, and they confirm that these links aren't prefetched until a hover event fires on them. I should also verify that that's all that happens!

## Docs

Probably needs docs? I'm happy to write them @withastro/maintainers-docs

### CODE REVIEW

**Comment by ematipico** [comment]
I believe this option should accept an array of selectors.

While it's true that most of the times developers should use simple anchor elements for links, sometimes there are edge cases where developers must use different elements, e.g. div, button, span, etc.

**Comment by kory-smith** [comment]
> I believe this option should accept an array of selectors.
> 
> While it's true that most of the times developers should use simple anchor elements for links, sometimes there are edge cases where developers must use different elements for whatever reason, e.g. div, button, span, etc.
> 
> By passing an array of selectors, we can give them more control over which links or link-like elements can prefetch.

In your opinion, should both the "visible" and "hover" prefetchers be refactored to accept an array? I figure it's more appropriate for both of them to accept arrays, but I feel that refactoring the behavior of the existing prefetch integration is out of scope for this PR.

**Comment by kory-smith** [comment]
> > > I believe this option should accept an array of selectors.
> > > While it's true that most of the times developers should use simple anchor elements for links, sometimes there are edge cases where developers must use different elements for whatever reason, e.g. div, button, span, etc.
> > > By passing an array of selectors, we can give them more control over which links or link-like elements can prefetch.
> > 
> > 
> > In your opinion, should both the "visible" and "hover" prefetchers be refactored to accept an array? I figure it's more appropriate for both of them to accept arrays, but I feel that refactoring the behavior of the existing prefetch integration is out of scope for this PR.
> 
> That's a good question. Unfortunately, I don't have a good answer for you, but I only have one question: what are the goals and non-goals of your PR?
> 
> What are the expectations you set when you opened the PR? Unfortunately, there's no [RFC](https://github.com/withastro/roadmap/discussions), and it gets difficult for us to understand where's the line between "what to implement now" and "what to implement tomorrow".

Okay, here is my take. 

I think that both the visible and hover prefetchers should accept either a string or an array. That would preserve backwards compatibility with the existing prefetch integration custom selector logic. I'm new to contributing to frameworks, so I'm not sure how important backwards compatibility is. I figure it's very important.

So, this PR adds intent-based prefetching. I'll come back to make a separate PR that adds the option to pass an array of selectors to the visible prefetch integration--the current one. That way, we preserve backwards compatibility and also enhance the integration.

If you agree with that stance, I already added the either-string-or-array logic and wrote tests for it. All that's left are docs.

**Review by matthewp — APPROVED** [review]
This makes sense to me

**Comment by kory-smith** [comment]
> This makes sense to me

Given your approval, what needs to happen in order to get this merged? Docs?

**Comment by Princesseuh** [comment]
Hello, sorry for the lack of answer. If you're still willing to work on this, docs would be the only thing missing yep! Just add a mention of this new setting in the "Configuration" section would be enough.

**Review by natemoo-re — APPROVED** [review]
I'm also good with this change, thanks so much for the PR!

cc-ing @withastro/maintainers-docs or @tony-sull for some help updating the README to get this over the line!

**Comment by sarah11918** [comment]
Happy to help with docs!

If this is an integration to be installed, then it would need a package README that would automatically show up here: https://docs.astro.build/en/guides/integrations-guide

Assuming there's not a lot to configure with this integration, I think the prefix integration might be a good model for docs: https://docs.astro.build/en/guides/integrations-guide/prefetch/ (source file: https://github.com/withastro/astro/blob/main/packages/integrations/prefetch/README.md )

We'd also need an svg logo for the integration, and to list it here https://github.com/withastro/docs/blob/main/src/data/logos.ts so it can be pulled into our integrations navigation component.

Is that enough to get started?

**Comment by kory-smith** [comment]
Hey, all. It's been a while! I'm happy to write the docs for this bad boy. I'll have a first draft done by, say, Friday the 16th? I'll post them here. Thanks all for the kind words!

**Comment by kory-smith** [comment]
Okay, friends. I posted the docs changes to the PR. For easy reviewing, here's what I changed (new stuff in bold)

## Usage

When you install the integration, the prefetch script is automatically added to every page in the project. Just add `rel="prefetch"` to any `<a />` links on your page and you're ready to go!

**In addition, you can add `rel="prefetch-intent"` to any `<a />` links on your page to prefetch them only when they are hovered over, touched, or focused. This is especially useful when you want to be conservative with your users' data usage.**

## Configuration

The Astro Prefetch integration handles which links on the site are prefetched and it has its own options. Change these in the `astro.config.mjs` file which is where your project's integration settings live.

### config.selector
  
By default the prefetch script searches the page for any links that include a `rel="prefetch"` attribute, ex: `<a rel="prefetch" />` or `<a rel="nofollow prefetch" />`. This behavior can be changed in your `astro.config.*` file to use a custom query selector when finding prefetch links. 

__`astro.config.mjs`__

```js
import { defineConfig } from 'astro/config';
import prefetch from '@astrojs/prefetch';

export default defineConfig({
  // ...
  integrations: [prefetch({
    // Only prefetch links with an href that begins with `/products`
    selector: "a[href^='/products']"
  })],
});
```

**### config.intentSelector
By default, the prefetch script also searches the page for any links that include a `rel="prefetch-intent"` attribute, ex: `<a rel="prefetch-intent" />`. This behavior can be changed in your astro.config.* file to use a custom query selector when finding prefetch-intent links.

Importantly, the intentSelector accepts both strings and arrays, giving you the option to prefetch multiple selectors on hover if you want.

__`astro.config.mjs`__

```js
import { defineConfig } from 'astro/config';
import prefetch from '@astrojs/prefetch';

export default defineConfig({
  // ...
  integrations: [prefetch({
    // Only prefetch links with an href that begins with `/products` or `/coupons`
    intentSelector: ["a[href^='/products']", "a[href^='/coupons']"]

    // Below is also valid syntax
    // intentSelector: "a[href^='/products']"
  })],
});
```**

### config.throttle
  
By default the prefetch script will only prefetch one link at a time. This behavior can be changed in your `astro.config.*` file to increase the limit for concurrent downloads.

__`astro.config.mjs`__

```js
import { defineConfig } from 'astro/config';
import prefetch from '@astrojs/prefetch';

export default defineConfig({
  // ...
  integrations: [prefetch({
    // Allow up to three links to be prefetched concurrently
    throttle: 3
  })],
});
```

## Troubleshooting
- If your installation doesn't seem to be working, try restarting the dev server.
- If a link doesn't seem to be prefetching, make sure that the link is pointing to a page on the same domain and matches the integration's `selector` option.

For help, check out the `#support` channel on [Discord](https://astro.build/chat). Our friendly Support Squad members are here to help!

You can also check our [Astro Integration Documentation][astro-integration] for more on integrations.

## Contributing

This package is maintained by Astro's Core team. You're welcome to submit an issue or PR!

## Changelog

See [CHANGELOG.md](CHANGELOG.md) for a history of changes to this integration.

[astro-integration]: https://docs.astro.build/en/guides/integrations-guide/

File: packages/integrations/prefetch/README.md [hunk 1]
```diff
@@ -78,6 +80,30 @@ export default defineConfig({
   })],
 });
 ```
+
+### config.intentSelector
+By default, the prefetch script also searches the page for any links that include a `rel="prefetch-intent"` attribute, ex: `<a rel="prefetch-intent" />`. This behavior can be changed in your astro.config.* file to use a custom query selector when finding prefetch-intent links.
+
+Importantly, the intentSelector accepts both strings and arrays, giving you the option to prefetch multiple selectors on hover if you want.
```
Comment (sarah11918) [inline packages/integrations/prefetch/README.md:87]:
```suggestion
```
I think we can handle this with strategic inline commenting instead...

File: packages/integrations/prefetch/README.md [hunk 2]
```diff
@@ -78,6 +80,30 @@ export default defineConfig({
...
+import prefetch from '@astrojs/prefetch';
+
+export default defineConfig({
+  // ...
+  integrations: [prefetch({
+    // Only prefetch links with an href that begins with `/products` or `/coupons`
+    intentSelector: ["a[href^='/products']", "a[href^='/coupons']"]
+
+    // Below is also valid syntax
+    // intentSelector: "a[href^='/products']"
```
Comment (sarah11918) [inline packages/integrations/prefetch/README.md:102]:
```suggestion
     // Only prefetch links with an href that begins with `/products` or `/coupons`
    intentSelector: ["a[href^='/products']", "a[href^='/coupons']"]
    
    // Use a string to prefetch a single selector
    // intentSelector: "a[href^='/products']"
```
I think this is a bit more illustrative than the previous, and means we can avoid a sentence!

**Review by sarah11918 — COMMENTED** [review]
What a cool feature, @kory-smith! Thanks for using the existing structure to prepare docs for this. I can tell you were careful to model after the rest of the page, and I appreciate that! I made a couple of eding suggestions for your consideration. See what you think!

**Comment by kory-smith** [comment]
> What a cool feature, @kory-smith! Thanks for using the existing structure to prepare docs for this. I can tell you were careful to model after the rest of the page, and I appreciate that! I made a couple of eding suggestions for your consideration. See what you think!

@sarah11918, I like all your suggestions quite a lot, and so I have committed them all. I definitely should have noticed that there was no "user" language in the existing docs and written mine accordingly. My bad!

**Comment by kory-smith** [comment]
@sarah11918, wanted to bother you again to see if those doc revisions pass muster. :)

File: packages/integrations/prefetch/README.md [hunk 3]
```diff
@@ -78,6 +80,29 @@ export default defineConfig({
   })],
 });
 ```
+
+### config.intentSelector
+By default, the prefetch script also searches the page for any links that include a `rel="prefetch-intent"` attribute, ex: `<a rel="prefetch-intent" />`. This behavior can be changed in your astro.config.* file to use a custom query selector when finding prefetch-intent links.
```
Comment (sarah11918) [inline packages/integrations/prefetch/README.md:85]:
```suggestion
By default, the prefetch script also searches the page for any links that include a `rel="prefetch-intent"` attribute, ex: `<a rel="prefetch-intent" />`. This behavior can be changed in your `astro.config.*` file to use a custom query selector when finding prefetch-intent links.
```
Never let me look at something again, or I'll find something else!  :joy:

Yes, good to go! Sorry, in general it's usually fine if you just applied my suggestions, and I don't need to review again. But, while I'm here, let's inline code the file name here!

**Review by sarah11918 — APPROVED** [review]
Docs approves after the quick filename syntax fix!  :partying_face:

**Pull request merged**